core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -437,29 +436,52 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
     val rlsMetadata = getRemoteLogSegmentMetadata(tp, offset)
 
     var startPos = lookupPositionForOffset(rlsMetadata, offset)
-    var is:InputStream = null
+    var remoteSegInputStream: InputStream = null
     try {
       // Search forward for the position of the last offset that is greater than or equal to the target offset
-      is = remoteLogStorageManager.fetchLogSegmentData(rlsMetadata, startPos, Int.MaxValue)
-      val logInputStream = new RemoteLogInputStream(is)
+      remoteSegInputStream = remoteLogStorageManager.fetchLogSegmentData(rlsMetadata, startPos, Int.MaxValue)
+      val remoteLogInputStream = new RemoteLogInputStream(remoteSegInputStream)
 
       var batch:RecordBatch = null;
-      while ((batch = logInputStream.nextBatch()) != null && batch.lastOffset < offset) {
-        startPos += batch.sizeInBytes
+      def fetchBatch() : RecordBatch = {
+        batch = remoteLogInputStream.nextBatch()
+        batch
       }
-    } finally {
-      is.close()
-    }
 
-    try {
-      is = remoteLogStorageManager.fetchLogSegmentData(rlsMetadata, startPos, startPos + maxBytes)
-      val buffer = ByteBuffer.allocate(fetchMaxBytes)
-      Utils.readFully(is, buffer)
+      // always look for the batch which has the desired offset and set the startPosition as that segments start
+      // location.
+      var lastBatchSize = 0
+
+      // we will always have a batch in that segment as it is a non-compacted topic. For compacted topics, we may need
+      // to read from the subsequent segments if there is no batch available for the desired offset in the current
+      //segment. Because immediately available offset later to the desired offset will be available in the next segment.
+      // We will add that logic when we support compacted topics for tiering.
+      while (fetchBatch() != null && batch.lastOffset < offset) {
```

Comment:
@ying-zheng Are there any other cases where we need to look forward in subsequent segment when the topic is non-compacted?

Reply:
We need to take care of `FetchIsolation` level too. I will handle that in this PR.

Reply:
The existing log segment fetch logic does look forward in the next segment. But I don't know if there is any case that the segment returned by getRemoteLogSegmentMetadata() does not contain the offset we are looking for.

To be safer, I think it's better to follow the log segment fetch logic: if the offset is not found in the current segment, look up the offset in the next segment.

It's not an easy job to support transactions in tiered storage. I don't think we can address that in this PR